llm-translator.html

```diff
@@ -0,0 +1,276 @@
+        window.call = DailyIframe.createCallObject({
+          url: ROOM_URL,
+          subscribeToTracksAutomatically: false,
+          audioSource: host ? micSelect.value : false,
+          videoSource: host ? camSelect.value : false,
+          dailyConfig: {},
+        });
+
+        [
+          "participant-updated",
+          "participant-left",
+          "participant-joined",
+          "error",
+          "camera-error",
+          "started-camera",
+          "active-speaker-change",
+          "recording-started",
+          "recording-stopped",
+          "recording-error",
+          "track-started",
+          "track-stopped",
+        ].forEach((e) => {
+          call.on(e, (ev) => {
+            console.log(`🔥 ${e}:`, ev);
+          });
+        });
+
+        call.on("track-started", startTrack);
+        call.on("track-stopped", stopTrack);
+
+        call.on("participant-joined", (ev) => {
+          const p = ev.participant;
+          // is this a person or a translator bot?
+
+          if (isTranslator(p)) {
+            // This is a translator. Update the list of languages
+            // to include this translator's language
+            const lang = translatorLanguage(p);
+            const newOpt = document.createElement("option");
+            newOpt.value = lang;
+            newOpt.text = cap(lang);
+            languageSelect.appendChild(newOpt);
+          } else {
+            // This is a human. Subscribe to their video, and their
+            // audio if the current language is English
+            const audioSub = languageSelect.value === "english";
+            call.updateParticipant(p["session_id"], {
+              setSubscribedTracks: {
+                video: true,
+                audio: audioSub,
+              },
+            });
+          }
+        });
+
+        call.on("app-message", (e) => {
+          const lang = languageSelect.value;
+          if (e.fromId === "transcription" && lang === "english") {
+            showSubtitle(e.data.session_id, e.data.text);
+          } else if (e.data.translation_language === lang) {
+            showSubtitle(e.data.session_id, e.data.translation);
+          }
+          e.fromId;
+        });
+        await call.join({});
+      }
+
+      async function leaveRoom() {
+        let vids = document.getElementsByTagName("video");
+        for (let vid of vids) {
+          vid.remove();
+        }
+        await call.leave();
+      }
+      async function loadDevices() {
+        const dummy = await navigator.mediaDevices.getUserMedia({
+          audio: true,
+          video: true,
+        });
+        let dummyTracks = dummy.getTracks();
+        for (t in dummyTracks) {
+          dummyTracks[t].stop();
+        }
+
+        var devices = await navigator.mediaDevices.enumerateDevices();
+        var cameras = devices.filter((d) => d.kind == "videoinput");
+        var mics = devices.filter((d) => d.kind == "audioinput");
```

Comment:
```suggestion
        const mics = devices.filter((d) => d.kind == "audioinput");
```

Aside: may be more efficient to just loop through the devices once, check their kinds, and append to relevant mic or cam device array (vs having these two `filter` ops)